### PR TITLE
[JENKINS-50568] make receive queue max threads configurable

### PR DESCRIPTION
**Pull request by mawinter69**
use a ThreadPoolExecutor to process the received events. This allows to
easily reconfigure the number of threads similar to the GerritSendQueue

### CODE REVIEW

**Comment by mawinter69** [comment]
For the Jenkins gerrit-trigger plugin, a change is also required to provide a ThreadPoolFactory that is jenkins aware (runs with Jenkins system ACL)

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritDefaultValues.java [hunk 1]
```diff
@@ -84,6 +84,10 @@ private GerritDefaultValues() {
      * The default nr of event worker threads.
      */
     public static final int DEFAULT_NR_OF_RECEIVING_WORKER_THREADS = 3;
+    /**
+     * The default nr of event worker threads.
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritDefaultValues.java]: Copy paste error?

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java [hunk 2]
```diff
@@ -84,33 +94,104 @@
...
+    /**
+     * Create handler with the given number of maximum worker threads and given thread keep alive
+     * time.
+     *
+     * @param numberOfWorkerThreads the number of event threads.
+     * @param threadKeepAliveTime the number of seconds threads will stay alive.
+     */
+    public GerritHandler(int numberOfWorkerThreads, int threadKeepAliveTime) {
         this.numberOfWorkerThreads = numberOfWorkerThreads;
+        this.threadKeepAliveTime = threadKeepAliveTime;
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: Since `executor.allowCoreThreadTimeOut(true);` is specified, I think we need to specify some minimum safe number here. According to the documentation setting for example the timeout to 0 will cause new threads to be created all the time and thrown away upon completion. And I think that might not be desirable.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java [hunk 3]
```diff
@@ -84,33 +94,104 @@
...
+                  executor.getPoolSize(), getQueueSize());
+          logger.info("Nr of active pool-threads: {}", executor.getActiveCount());
+      }
+    }
+
+    /**
+     * Returns the largest number of threads that have ever simultaneously been in the pool.
+     * @return number of threads
+     */
+    int getLargestPoolSize() {
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: why package private?

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java [hunk 4]
```diff
@@ -386,10 +540,26 @@ private boolean ignoreEvent(CommentAdded event) {
...
     public void shutdown(boolean join) {
-        for (EventThread worker : workers) {
-            worker.shutdown();
+      if (executor != null) {
+        ThreadPoolExecutor pool = executor;
+        executor = null;
+        pool.shutdown(); // Disable new tasks from being submitted
+        try {
+            // Wait a while for existing tasks to terminate
+            if (!pool.awaitTermination(WAIT_FOR_JOBS_SHUTDOWN_TIMEOUT, TimeUnit.SECONDS)) {
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: You should only `awaitTermination` if `join == true`

File: src/test/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandlerQueueTest.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+     */
+    @Test
+    public void testDefaultNumberofParallelWorkers() throws Exception {
+        handler = new GerritHandler();
+
+        listener = new SlowEventListener();
+        handler.addListener(listener);
+        postEventsToQueue(5);
+        waitForEventsProcessed();
+        assertThat(handler.getLargestPoolSize(), equalTo(3));
```
Comment (rsandell) [inline src/test/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandlerQueueTest.java]: Better to use the constant here, so an onlooker understands it tests the default number

File: src/test/java/com/sonymobile/tools/gerrit/gerritevents/SSHUtilTest.java [hunk 6]
```diff
@@ -24,9 +23,7 @@
     @Test
     public void testNoPassphraseParsing() throws Exception {
         // Get no-passphrase key resource as file
-        URL url = Thread.currentThread().getContextClassLoader().getResource(
-                "com/sonymobile/tools/gerrit/gerritevents/id_rsa");
-        File file = new File(url.getPath());
+        File file = new File("src/test/resources/com/sonymobile/tools/gerrit/gerritevents/id_rsa");
```
Comment (rsandell) [inline src/test/java/com/sonymobile/tools/gerrit/gerritevents/SSHUtilTest.java]: Why did you change this?
this won't work when running the tests from the published test jar

File: src/test/java/com/sonymobile/tools/gerrit/gerritevents/workers/EventThreadTest.java [hunk 7]
```diff
@@ -77,6 +78,7 @@ public static void beforeClass() {
      * @throws Exception if so.
      */
     @Test(timeout = 1000)
+    @Ignore
```
Comment (rsandell) [inline src/test/java/com/sonymobile/tools/gerrit/gerritevents/workers/EventThreadTest.java]: Probably better to just remove these tests

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java [hunk 8]
```diff
@@ -183,6 +276,7 @@ private void post(Work work) {
         try {
             logger.trace("putting work on queue.");
             workQueue.put(work);
+            queueWork(work);
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: Why double storage of the events? You seem to have removed most of the `workQueue` usage already.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java (hunk 2 above)
Comment (mawinter69) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: what would be a reasonable minimum value? 30 seconds? Maybe we could also use a fixed value (not allowing it to be changed but that makes testing harder and when you reduce the number of threads it might take quite long until they die.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java (hunk 3 above)
Comment (mawinter69) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: Does it have any value for users to know what is the max poolsize that was encountered? I need this method for testing to be at least package private.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java (hunk 8 above)
Comment (mawinter69) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: I will remove this completely, but then the getWorkQueue method should also be removed which is an API change.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java (hunk 4 above)
Comment (mawinter69) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: done

File: src/test/java/com/sonymobile/tools/gerrit/gerritevents/SSHUtilTest.java (hunk 6 above)
Comment (mawinter69) [reply src/test/java/com/sonymobile/tools/gerrit/gerritevents/SSHUtilTest.java]: ok, but with the url it doesn't work with maven (at least on windows)

File: src/test/java/com/sonymobile/tools/gerrit/gerritevents/workers/EventThreadTest.java (hunk 7 above)
Comment (mawinter69) [reply src/test/java/com/sonymobile/tools/gerrit/gerritevents/workers/EventThreadTest.java]: gone

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritDefaultValues.java (hunk 1 above)
Comment (mawinter69) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritDefaultValues.java]: fixed

File: src/test/java/com/sonymobile/tools/gerrit/gerritevents/SSHUtilTest.java (hunk 6 above)
Comment (rsandell) [reply src/test/java/com/sonymobile/tools/gerrit/gerritevents/SSHUtilTest.java]: Yea, I would guess this is full of test code not working on windows :) Though still a bit strange why this wouldn't work.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java (hunk 3 above)
Comment (rsandell) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: ok, in those cases a comment explaining it's there for testability is common :)

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java (hunk 8 above)
Comment (rsandell) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: `executor.getQueue()` ?

Comment (mawinter69) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: executor.getQueue() returns BlockingQueue<Runnable> while  the interface defines BlockingQueue<Work>

Comment (rsandell) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: You could then keep the `workQueue` reference to the queue you give the executor. They would then be the same instance but with the correct type. Or just cast it since we control what goes in it I guess, both ways are slightly ugly in their own ways :)

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java [hunk 9]
```diff
@@ -84,35 +94,101 @@
...
+     * Create handler with the given number of maximum worker threads and given thread keep alive
+     * time (not allowing the thread keep alive to become less than 
+     * @see GerritDefaultValues#MIN_RECEIVE_THREAD_KEEP_ALIVE_TIME)
+     *
+     * @param numberOfWorkerThreads the number of event threads.
+     * @param threadKeepAliveTime the number of seconds threads will stay alive. 
+     */
+    public GerritHandler(int numberOfWorkerThreads, int threadKeepAliveTime) {
+        this.numberOfWorkerThreads = numberOfWorkerThreads;
+        if (threadKeepAliveTime < MIN_RECEIVE_THREAD_KEEP_ALIVE_TIME)
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: `Math.max`

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/workers/Coordinator.java [hunk 10]
```diff
@@ -32,12 +32,6 @@
  * @author Robert Sandell &lt;robert.sandell@sonyericsson.com&gt;
  */
 public interface Coordinator {
-    /**
-     * Retrieves the work queue for workers to poll.
-     * @return the queue
-     */
-    BlockingQueue<Work> getWorkQueue();
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/workers/Coordinator.java]: No, unfortunately we can't change binary signature like this. Something else needs to be done.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/workers/EventThread.java [hunk 11]
```diff
@@ -1,81 +0,0 @@
...
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-/**
- * A basic event thread.
- * The idea is to split up as much work as possible to be able to quickly handle the next event from Gerrit
- * @author Robert Sandell &lt;robert.sandell@sonyericsson.com&gt;
- */
-public class EventThread extends Thread {
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/workers/EventThread.java]: Speaking of binary compatibility, this needs to stay but be deprecated.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java (hunk 2 above)
Comment (rsandell) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: I think we can go lower, even 10 seconds. IIUC as long as it's not `0` the core thread count should stay the same.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java (hunk 8 above)
Comment (mawinter69) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: Maybe we can make Work implement also Runnable.
Then we could cast `executor.getQueue()` to `BlockingQueue<Work>`

Comment (rsandell) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: Not very easy without breaking more compatibility stuff, the way it's designed with the 'perform(Coordinator)' method.
But the simple fix to keep it would be something like
```java
List<Work> l = new ArrayList<>();
for (Runnable r : executor.getQueue()) {
    if (r instanceof EventWorker) {
        l.add(((EventWorker)l).work);
    }
}
```
I think that would iterate normally without blocking, but needs testing.

There might be some callers that are expecting to get the reference to the actual queue, so the javadoc needs to be updated as well. But at least we keep binary compatibility and hopefully the deprecation warning will make it so no user uses it any more.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java [hunk 12]
```diff
@@ -289,12 +406,45 @@ public int getNumberOfWorkerThreads() {
...
+    public int getThreadKeepAliveTime() {
+        return threadKeepAliveTime;
+    }
+
+    /**
+     * Sets the number of seconds threads well be kept alive.
+     *
+     * @param threadKeepAliveTime number of seconds
+     */
+    public void setThreadKeepAliveTime(int threadKeepAliveTime) {
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: Needs same `MIN_RECEIVE_THREAD_KEEP_ALIVE_TIME` guard as in the constructor

Comment (rsandell) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: also check for `null`

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java [hunk 13]
```diff
@@ -289,12 +406,45 @@ public int getNumberOfWorkerThreads() {
...
+     * Returns a snapshot of the current state of the queue in the executor and not the Queue itself.
+     * 
+     * {@inheritDoc}
+     */
     @Override
     public BlockingQueue<Work> getWorkQueue() {
-        return workQueue;
+
+        BlockingQueue<Work> queue = new LinkedBlockingQueue<Work>();
+        BlockingQueue<Runnable> workQueue = executor.getQueue();
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: check for `null`

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java [hunk 14]
```diff
@@ -84,35 +94,97 @@
...
+    }
+
+    /**
+     * Returns the largest number of threads that have ever simultaneously been in the pool.
+     * Package visibility for testing purposes only.
+     *
+     * @return number of threads
+     */
+    int getLargestPoolSize() {
+        return executor.getLargestPoolSize();
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: check for `null`

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java [hunk 15]
```diff
@@ -386,10 +536,28 @@ private boolean ignoreEvent(CommentAdded event) {
      * @param join if the method should wait for the thread to finish before returning.
      */
     public void shutdown(boolean join) {
-        for (EventThread worker : workers) {
-            worker.shutdown();
+        if (executor != null) {
+            ThreadPoolExecutor pool = executor;
+            executor = null;
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: there are places throughout in the code that are assuming a non null executor. See above comments but I have most likely not found them all.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/workers/Coordinator.java [hunk 16]
```diff
@@ -28,14 +28,15 @@
...
- * Representation interface of an EventThread workers coordinator.
+ * Representation interface of a Work coordinator.
  * @author Robert Sandell &lt;robert.sandell@sonyericsson.com&gt;
  */
 public interface Coordinator {
     /**
      * Retrieves the work queue for workers to poll.
      * @return the queue
      */
+    @Deprecated
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/workers/Coordinator.java]: Needs a walkthrough of gerrit-trigger-plugin to figure out what it needed this for and probably add whatever it actually needs. A bit iffy in terms of binary compatibility. It might be time to go to java 8 and add default methods, or introduce a `Coordinator2` interface?

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java [hunk 17]
```diff
@@ -84,35 +94,97 @@
...
     /**
-     * Create the Event Thread.
-     * @param threadName Name of thread to be created.
-     * @return new EventThread to be used by worker
+     * Returns a factory for creating worker threads for the receiving queue.
+     * Overwrite when you have special requirements, e.g. security constraints
+     *
+     * @return ThreadFactory
      */
-    protected EventThread createEventThread(String threadName) {
```
Comment (rsandell) [inline src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: oh, was this removed as well, sorry I missed it. Same binary compatibility problems as the other removed methods I think.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java (hunk 15 above)
Comment (mawinter69) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java]: actually executor can't be null. It is initialized in the constructor.

File: src/main/java/com/sonymobile/tools/gerrit/gerritevents/workers/Coordinator.java (hunk 16 above)
Comment (mawinter69) [reply src/main/java/com/sonymobile/tools/gerrit/gerritevents/workers/Coordinator.java]: Gerrit trigger plugin is not using the `getWorkQueue()` method.
